### PR TITLE
Add PolygonDataQueueHandler

### LINKED ISSUE #3380: Create an IQueueDataHandler / IDataDownloader implementation for Polygon API.
#### Expected Behavior
[Polygon.io](https://polygon.io/docs/#getting-started) provides affordable market data via API. It would be nice to have this supported in LEAN. Their API can support historical download and live data.

#### Actual Behavior
No Polygon implementation.

#### Potential Solution
Implement Polygon.

#### Checklist
- [x] I have completely filled out this template
- [x] I have confirmed that this issue exists on the current `master` branch
- [x] I have confirmed that this is not a duplicate issue by searching [issues](https://github.com/QuantConnect/Lean/issues)

### PR DESCRIPTION
**Pull request by StefanoRaggi**
#### Description
- Added a new `PolygonDataQueueHandler` with support for `Equity`, `Forex` and `Crypto` security types.
  - Implements `IDataQueueHandler` and `IHistoryProvider`
- Added a new `PolygonDataDownloader`

#### Related Issue
Closes #3380

#### Motivation and Context
- Additional data provider

#### Requires Documentation Change
No.

#### How Has This Been Tested?
- Local testing with live algorithm

#### Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:
- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. 
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

File: ToolBox/Polygon/PolygonDataDownloader.cs [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+            if (resolution != Resolution.Minute &&
+                resolution != Resolution.Hour &&
+                resolution != Resolution.Daily)
+            {
+                throw new NotSupportedException($"Resolution not supported: {resolution}");
+            }
+
+            if (symbol.SecurityType != SecurityType.Equity)
+            {
+                throw new NotSupportedException($"Security type not supported: {symbol.SecurityType}");
```
Comment (Martin-Molinero) [inline ToolBox/Polygon/PolygonDataDownloader.cs:51]: I believe polygon does also have crypto/forex historical trade data?

File: ToolBox/Polygon/PolygonWebSocketClientWrapper.cs [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+            Open += OnOpen;
+            Closed += OnClosed;
+            Message += OnMessage;
+            Error += OnError;
+
+            Connect();
+        }
+
+        public void Subscribe(Symbol symbol, bool subscribe)
```
Comment (Martin-Molinero) [inline ToolBox/Polygon/PolygonWebSocketClientWrapper.cs:50]: Maybe we should have an `Unsubscribe` public method too + an internal private method with the `bool subscribe`

File: ToolBox/Polygon/PolygonDataDownloader.cs (hunk 1 above)
Comment (StefanoRaggi) [reply ToolBox/Polygon/PolygonDataDownloader.cs:53]: Done -- Minute/Hour resolution for Forex/Crypto is only available up to 5/25/2020 for now though

File: ToolBox/Polygon/PolygonDownloaderProgram.cs [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+                Console.WriteLine("--tickers=eg SPY,AAPL");
+                Console.WriteLine("--resolution=Minute/Hour/Daily");
+                Environment.Exit(1);
+            }
+
+            try
+            {
+                // Load settings from command line
+                var resolution = (Resolution)Enum.Parse(typeof(Resolution), resolutionString);
+                var securityType = SecurityType.Equity;
```
Comment (Martin-Molinero) [inline ToolBox/Polygon/PolygonDownloaderProgram.cs:43]: I think Security type should be an argument now too? And the market probably also?

File: ToolBox/Polygon/PolygonDataQueueHandler.cs [hunk 4]
```diff
@@ -0,0 +1,536 @@
...
+                {
+                    var utcTime = Time.UnixMillisecondTimeStampToDateTime(row.Timestamp);
+
+                    if (utcTime < request.StartTimeUtc ||
+                        utcTime > request.EndTimeUtc.Add(request.Resolution.ToTimeSpan()))
+                    {
+                        continue;
+                    }
+
+                    var time = utcTime.ConvertFromUtc(TimeZones.NewYork);
```
Comment (Martin-Molinero) [inline ToolBox/Polygon/PolygonDataQueueHandler.cs:303]: TimeZone should be exchange I think here? Thinking about Crypto/Forex for example. And the `PolygonDataDownloader` should be using Data time zone for storing the data, right?
Just checking, for forex we only have quote streamed data, but history are trades?

Comment (Martin-Molinero) [reply ToolBox/Polygon/PolygonDataQueueHandler.cs:303]: I think there seems to be NBBO quotes history https://polygon.io/docs/#get_v2_ticks_stocks_nbbo__ticker___date__anchor?

**Pull request merged**